Source/Core/Duality/Utility/Coroutines/Coroutine.cs

```diff
@@ -0,0 +1,116 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Duality.Resources;
+
+namespace Duality.Utility.Coroutines
+{
+	/// <summary>
+	/// A Coroutine's management class
+	/// </summary>
+	public sealed class Coroutine : IDisposable
+	{
+		private IEnumerator<WaitUntil> enumerator = null;
+		private WaitUntil currentCondition;
+
+		/// <summary>
+		/// The Coroutine's current <see cref="CoroutineStatus">execution status</see>
+		/// </summary>
+		public CoroutineStatus Status { get; private set; }
+
+		/// <summary>
+		/// The Exception encountered during the last frame
+		/// </summary>
+		public Exception LastException { get; private set; }
+
+		/// <summary>
+		/// The Coroutine's Name
+		/// </summary>
+		public string Name { get; private set; }
+
+		/// <summary>
+		/// True if this Coroutine's Status is <see cref="CoroutineStatus.Paused"/> or <see cref="CoroutineStatus.Running"/>
+		/// </summary>
+		public bool IsAlive
+		{
+			get { return this.Status == CoroutineStatus.Paused || this.Status == CoroutineStatus.Running; }
+		}
+
+		internal void Setup(IEnumerable<WaitUntil> values, string name)
+		{
+			this.Status = CoroutineStatus.Running;
+
+			this.LastException = null;
+			this.Name = name;
+
+			if (this.enumerator != null)
+				this.enumerator.Dispose();
+
+			this.enumerator = values.GetEnumerator();
+
+			this.MoveNext();
+		}
+
+		internal void Update()
+		{
+			if (this.Status != CoroutineStatus.Running) return;
+
+			try
+			{
+				if (this.currentCondition.HasEndedAfterUpdate())
+					this.MoveNext();
+			}
+			catch (Exception e)
+			{
+				Logs.Core.WriteError("An error occurred while processing Coroutine '{0}': {1}", this.Name, LogFormat.Exception(e));
+
+				this.LastException = e;
+				this.Status = CoroutineStatus.Error;
+			}
+		}
+
+		private void MoveNext()
```

Comment:
The naming is a bit confusing as it implies this implements IEnumerator. Its also very small so I think it might be more readable to remove this.

Reply:
but the logic is used in two different parts, hence the separate method

Reply:
Looking at this code a bit more it seems that when you call StartCoroutine it immediately executes the first part of the coroutine. I think this should be done when the update method is called. This is more in line with how IEnumerator normally functions. To give a example your unit test will change into this:
```cs
		[Test] public void Basics()
		{
			Scene scene = new Scene();
			scene.Activate();

			CoroutineObject obj = new CoroutineObject();
			Coroutine coroutine = scene.StartCoroutine(this.BasicRoutine(obj), "basic");

			scene.Update(); // This is new

			// All code until first yield is already executed
			Assert.AreEqual(10, obj.Value);
			Assert.True(coroutine.Status == CoroutineStatus.Running);

			scene.Update();

			// All code until second yield is executed
			Assert.AreEqual(20, obj.Value);
			Assert.True(coroutine.Status == CoroutineStatus.Running);

			scene.Update();

			// Yelded null, value didn't change, need to wait one more update
			Assert.AreEqual(20, obj.Value);
			Assert.True(coroutine.Status == CoroutineStatus.Running);

			scene.Update();
			// Yielded condition is waiting for two frames now..
			Assert.AreEqual(20, obj.Value);
			Assert.True(coroutine.Status == CoroutineStatus.Running);

			scene.Update();
			// All remaining code has been executed
			Assert.AreEqual(30, obj.Value);
			Assert.True(coroutine.Status == CoroutineStatus.Complete);

			scene.Update();

			// No further changes
			Assert.AreEqual(30, obj.Value);
			Assert.True(coroutine.Status == CoroutineStatus.Complete);
		}
```

With the addition of the update call the unit test are still green,

The coroutine Update method now looks like this (still with a field because we are basically using a reference to a struct but we can probably write it in a way this is not needed):
```cs
		internal void Update()
		{
			if (this.Status != CoroutineStatus.Running) return;

			try
			{
				if (this.currentCondition.HasEndedAfterUpdate())
				{
					if (this.enumerator.MoveNext())
						this.currentCondition = this.enumerator.Current;
					else
						this.Status = CoroutineStatus.Complete;
				}

			}
			catch (Exception e)
			{
				Logs.Core.WriteError("An error occurred while processing Coroutine '{0}': {1}", this.Name, LogFormat.Exception(e));

				this.LastException = e;
				this.Status = CoroutineStatus.Error;
			}
		}
```

Reply:
> I think this should be done when the update method is called.

But the first time there would be no`currentCondition` set yet, this would introduce a further "if" at the start of the Update method making it more complex, and potentially having 2 blocks of code (before and after a one-frame wait) executed in the same frame.

Reply:
There will be a `currentCondition` but the default is now one that already ended